portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -29,26 +33,46 @@
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
 
-  ArrayList<String> sampleMessages = new ArrayList<>();
-
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {  
-    String messages = convertToJsonString(sampleMessages);
-    response.setContentType("text/html;");
-    response.getWriter().println(messages);
+    Query query = new Query("Comment").addSort("timestamp", SortDirection.DESCENDING);
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery results = datastore.prepare(query);
+    
+    ArrayList<Comment> comments = new ArrayList<>();
+    for(Entity entity : results.asIterable()) {
+        long id = entity.getKey().getId();
+        long timestamp = (long) entity.getProperty("timestamp");
+        String comment = (String) entity.getProperty("comment");
+
+        Comment userComment = new Comment(id, comment, timestamp);
+        comments.add(userComment);
+    }
+    String jsonComments = new Gson().toJson(comments);
+    response.setContentType("application/json;");
+    response.getWriter().println(jsonComments);
   }
 
   @Override
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
     String comment = getUserData(request, "user-comment");
     String name = getUserData(request, "user-name");
+
     if(comment == null || name == null) {
       response.setContentType("text/html");
       response.getWriter().println("Please make sure both entries are filled.");
       return;
     }
     String userComment = name + " says " + comment;
-    sampleMessages.add(userComment);
+    long timestamp = System.currentTimeMillis();
+    
+    Entity taskEntity = new Entity("Comment");
```

Comment:
Just curious. Is this a JSON datastore where you don't need to define a schema?

Reply:
i am not very familiar with databases and schema as such. However, datastore is NoSQL database (if that info is helpful)

Reply:
Yeah thats what I thought. Thanks.